BrainPortal/lib/authenticated_system.rb

```diff
@@ -4,22 +4,36 @@ module AuthenticatedSystem #:nodoc:
   Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
 
   protected
+
+    # Extract the API token from the params, leaves it in @cbrain_api_token
+    def extract_api_token
+      @cbrain_api_token = params.delete(:cbrain_api_token).presence
+      true
+    end
+
     # Returns true or false if the user is logged in.
     # Preloads @current_user with the user model if they're logged in.
     def logged_in?
       !!current_user
     end
 
-    # Accesses the current user from the session.
-    # Future calls avoid the database because nil is not equal to false.
+    # Returns the currently logged-in user, if any.
     def current_user
-      @current_user ||= login_from_session unless @current_user == false
+      return @current_user if @current_user.present?
+      return nil if @current_user == false # false means we previously explicitely set current_user=(nil)
```

Comment:
I did not understand the comment. Can you explain?

Reply:
This method, surprisingly, is the main entry point to test whether or not a user is logged in. Tarek had created it such that when first invoked, it will try to identify the user ID from the session information. Initially the internal variable `@current_user` is set to nil. If it could find user from the session, it stored and cached it in`@current_user` to make further lookups fast. However, if it couldn't then it sets the internal variable to `false` as a way to both indicate there is no user, but also further calls to the method will not try again to resolve the user.

I kept the same semantics, except now we have two ways to find out which user is logged in: by the API token, or by the `session[:user_id]` (as it was).

Reply:
It's annoying that this system is full of side-effects (the false is assigned in `current_user=()` when passed nil in argument.